File: src/main/kotlin/com/nx3games/unrealpasser/protocol/UnrealPasserBackendModel.kt
package com.nx3games.unrealpasser.protocol

import com.jetbrains.rd.framework.FrameworkMarshallers
import com.jetbrains.rd.framework.IRdCall
import com.jetbrains.rd.framework.ISerializers
import com.jetbrains.rd.framework.base.ISerializersOwner
import com.jetbrains.rd.framework.base.RdExtBase
import com.jetbrains.rd.framework.impl.RdCall
import com.jetbrains.rd.ide.model.Solution

class UnrealPasserBackendModel(
    private val _execute: RdCall<String, String> = RdCall(
        FrameworkMarshallers.String,
        FrameworkMarshallers.String,
    ),
) : RdExtBase() {
    val execute: IRdCall<String, String>
        get() = _execute

    override val serializersOwner: ISerializersOwner
        get() = Companion

    override val serializationHash: Long
        get() = SERIALIZATION_HASH

    init {
        bindableChildren.add("execute" to _execute)
    }

    companion object : ISerializersOwner {
        const val EXTENSION_NAME: String = "UnrealPasserBackendModel"
        const val SERIALIZATION_HASH: Long = 0x5EA2_0000_0000_0001L

        override fun register(serializers: ISerializers) {
            registerSerializersCore(serializers)
        }

        override fun registerSerializersCore(serializers: ISerializers) {
        }
    }
}

fun Solution.getUnrealPasserBackendModel(): UnrealPasserBackendModel =
    getOrCreateExtension(UnrealPasserBackendModel.EXTENSION_NAME) { UnrealPasserBackendModel() }

File: backend/UnrealPasser.Backend/UnrealPasserBackendModel.cs
#nullable disable

using System;
using System.Collections.Generic;
using JetBrains.Rd;
using JetBrains.Rd.Base;
using JetBrains.Rd.Impl;
using JetBrains.Rd.Tasks;
using JetBrains.Rider.Model;

namespace UnrealPasser.Backend;

public sealed class UnrealPasserBackendModel : RdExtBase
{
    public const string ExtensionName = "UnrealPasserBackendModel";
    public const long ModelSerializationHash = 0x5EA2000000000001L;

    private readonly RdCall<string, string> myExecute;

    public UnrealPasserBackendModel()
        : this(new RdCall<string, string>(
            Serializers.ReadString,
            Serializers.WriteString,
            Serializers.ReadString,
            Serializers.WriteString))
    {
    }

    private UnrealPasserBackendModel(RdCall<string, string> execute)
    {
        myExecute = execute ?? throw new ArgumentNullException(nameof(execute));
        BindableChildren.Add(new KeyValuePair<string, object>("execute", myExecute));
    }

    public IRdEndpoint<string, string> Execute => myExecute;

    protected override long SerializationHash => ModelSerializationHash;

    protected override Action<ISerializers> Register => RegisterDeclaredTypesSerializers;

    private static void RegisterDeclaredTypesSerializers(ISerializers serializers)
    {
    }
}

public static class SolutionUnrealPasserBackendModelEx
{
    public static UnrealPasserBackendModel GetUnrealPasserBackendModel(this Solution solution)
    {
        return solution.GetOrCreateExtension(
            UnrealPasserBackendModel.ExtensionName,
            static () => new UnrealPasserBackendModel());
    }
}
